libp2p/autotls/autotls.nim

```diff
@@ -0,0 +1,258 @@
+# Nim-Libp2p
+# Copyright (c) 2025 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+{.push raises: [].}
+{.push public.}
+
+import
+  net,
+  results,
+  chronos,
+  chronicles,
+  bearssl/rand,
+  bio,
+  json,
+  chronos/apps/http/httpclient
+
+import ./acme
+import ./peeridauth
+import ./utils
+import ../nameresolving/dnsresolver
+import ../wire
+import ../crypto/crypto
+import ../peerinfo
+import ../utils/heartbeat
+
+logScope:
+  topics = "libp2p autotls"
+
+const
+  DefaultDnsServers =
+    @[
+      initTAddress("1.1.1.1:53"),
+      initTAddress("1.0.0.1:53"),
+      initTAddress("[2606:4700:4700::1111]:53"),
+    ]
+  DefaultRenewCheckTime = 1.hours
+  DefaultRenewBufferTime = 1.hours
+
+type SigParam = object
+  k: string
+  v: seq[byte]
+
+type AutoTLSManager* = ref object
+  rng: ref HmacDrbgContext
+  managerLoop: Future[void]
+  cert*: Opt[TLSCertificate]
+  certExpiry*: Opt[Moment]
+  certReady*: AsyncEvent
+  acmeAccount*: ref ACMEAccount
+  httpSession: HttpSessionRef
+  dnsResolver*: DnsResolver
+  peerInfo*: Opt[PeerInfo]
+  bearerToken*: Opt[string]
+  renewCheckTime*: Duration
+  renewBufferTime*: Duration
+  acmeServerURL: string
+
+proc checkDNSRecords(
+    self: AutoTLSManager, ip4Domain: string, acmeChalDomain: string, retries: int = 5
+): Future[bool] {.async: (raises: [AutoTLSError, CancelledError]).} =
+  var txt: seq[string]
+  var ip4: seq[TransportAddress]
+
+  for _ in 0 .. retries:
+    txt = await self.dnsResolver.resolveTxt(acmeChalDomain)
+    try:
+      ip4 = await self.dnsResolver.resolveIp(ip4Domain, 0.Port)
+    except:
+      raise newException(AutoTLSError, "Failed to resolve IP")
```

Comment:
Should we also use the retries for dns resolution failures? they do tend to happen (not frequently tho)